apps/emqx_management/src/emqx_mgmt_api_nodes.erl

```diff
@@ -250,55 +247,46 @@ nodes(get, _Params) ->
     list_nodes(#{}).
 
 node(get, #{bindings := #{node := NodeName}}) ->
-    get_node(NodeName).
+    emqx_api_lib:with_node(NodeName, to_ok_result_fun(fun get_node/1)).
 
 node_metrics(get, #{bindings := #{node := NodeName}}) ->
-    get_metrics(NodeName).
+    emqx_api_lib:with_node(NodeName, to_ok_result_fun(fun emqx_mgmt:get_metrics/1)).
 
 node_stats(get, #{bindings := #{node := NodeName}}) ->
-    get_stats(NodeName).
+    emqx_api_lib:with_node(NodeName, to_ok_result_fun(fun emqx_mgmt:get_stats/1)).
 
 %%--------------------------------------------------------------------
 %% api apply
 
 list_nodes(#{}) ->
-    NodesInfo = [format(Node, NodeInfo) || {Node, NodeInfo} <- emqx_mgmt:list_nodes()],
+    NodesInfo = [format(NodeInfo) || {_Node, NodeInfo} <- emqx_mgmt:list_nodes()],
     {200, NodesInfo}.
 
 get_node(Node) ->
-    case emqx_mgmt:lookup_node(Node) of
-        {error, _} ->
-            {400, #{code => 'SOURCE_ERROR', message => <<"rpc_failed">>}};
-        NodeInfo ->
-            {200, format(Node, NodeInfo)}
-    end.
-
-get_metrics(Node) ->
-    case emqx_mgmt:get_metrics(Node) of
-        {error, _} ->
-            {400, #{code => 'SOURCE_ERROR', message => <<"rpc_failed">>}};
-        Metrics ->
-            {200, Metrics}
-    end.
-
-get_stats(Node) ->
-    case emqx_mgmt:get_stats(Node) of
-        {error, _} ->
-            {400, #{code => 'SOURCE_ERROR', message => <<"rpc_failed">>}};
-        Stats ->
-            {200, Stats}
-    end.
+    format(emqx_mgmt:lookup_node(Node)).
 
 %%--------------------------------------------------------------------
 %% internal function
 
-format(_Node, Info = #{memory_total := Total, memory_used := Used}) ->
+format(Info = #{memory_total := Total, memory_used := Used}) ->
     Info#{
         memory_total := emqx_mgmt_util:kmg(Total),
         memory_used := emqx_mgmt_util:kmg(Used)
     };
-format(_Node, Info) when is_map(Info) ->
+format(Info) when is_map(Info) ->
     Info.
 
-node_error() ->
-    emqx_dashboard_swagger:error_codes([?SOURCE_ERROR], <<"Node error">>).
+to_ok_result({error, _} = Error) ->
+    Error;
+to_ok_result({ok, _} = Ok) ->
+    Ok;
+to_ok_result(Result) ->
+    {ok, Result}.
+
+to_ok_result_fun(Fun) when is_function(Fun) ->
```

Comment:
just a note to say, because of the hot-upgrade(I know, we haven't supported this feature) we rarely use an anonymous function, instead with the {M,F,A} tuple.

Reply:
Taken as is, but I'll keep it in mind for a future refactoring. Good point.